### PR TITLE
Add iterator implementation to BresenhamLine

### PR DESCRIPTION
**Pull request by sensualcoder**
Relevant to libtcod/libtcod#92

### CODE REVIEW

**Comment by HexDecimal** [comment]
Be careful with closing keywords. This pull request is to narrow to fully resolve that issue. You'd still want to mention the issue, just without using a special keyword.

File: src/libtcod/bresenham.h [hunk 1]
```diff
@@ -108,6 +112,26 @@ class TCODLIB_API BresenhamLine {
...
+    inline bool operator==(const iterator& rhs) const { return cur_ == rhs.cur_; }
+    inline bool operator!=(const iterator& rhs) const { return cur_ != rhs.cur_; }
+
+   private:
+    value_type cur_;
+    TCOD_bresenham_data_t& data_;
+  };
+
+  iterator begin() { return iterator({data_.origx, data_.origy}, data_); }
+  iterator end() { return iterator({data_.destx, data_.desty}, data_); }
```
Comment (HexDecimal) [inline src/libtcod/bresenham.h:133]: Keep in mind `end` in C++ means the past-the-end iterator.  The current implementation here will clip the endpoint.

File: src/libtcod/bresenham.h [hunk 2]
```diff
@@ -95,11 +96,14 @@ namespace tcod {
     This class is provisional.
  */
 class TCODLIB_API BresenhamLine {
+  using cell = std::pair<int, int>;
```
Comment (HexDecimal) [inline src/libtcod/bresenham.h:99]: I'd like to use `std::array<int, 2>` and then index the variable with `cur_.at(n)` for now.  The typename should describe that this is a 2 item position.

**Review by HexDecimal — CHANGES_REQUESTED** [review]
Bresenham itself should support a bidirectional iterator.  Relying on the C function is putting a limit on what this class is capable of doing.

When using this class as a range for line drawing you might need control whether or not the endpoints are included.  I'd suggest an adjust method that can tweak the position of the begin and end points forwards/backwards.

File: src/libtcod/bresenham.h (hunk 2 above)
Comment (sensualcoder) [reply src/libtcod/bresenham.h:99]: Would something like `xy_pos` work for this?

`using xy_pos = std::array<int, 2>;`

Comment (HexDecimal) [reply src/libtcod/bresenham.h:99]: That would work.  I tend to name custom typenames as if they're classes.
```suggestion
  using Point2 = std::array<int, 2>;
```

**Comment by sensualcoder** [comment]
Would it make sense to change the class to build the line list when the object is constructed, and then the iterator can then just reference that list of items? Something like this maybe:

```cpp
class TCODLIB_API BresenhamLine {
  using Point2 = std::array<int, 2>;

 public:
  BresenhamLine(Point2 from, Point2 to) {
    TCOD_bresenham_data_t data;
    TCOD_line_init_mt(from.at(0), from.at(1), to.at(0), to.at(1), &data);
    do {
      data_.push_back(from);
    } while (!TCOD_line_step_mt(&from.at(0), &from.at(1), &data));
  }

// ...

 private:
  std::size_t cur_ = 0;
  std::vector<Point2> data_{};
}
```

**Comment by HexDecimal** [comment]
> Would it make sense to change the class to build the line list when the object is constructed, and then the iterator can then just reference that list of items?

No, in that case you'd just return the `std::vector<Point2>` object which already holds the points and can be iterated over directly.  Holding that vector in the class is an unnecessary encapsulation.  `std::vector` also requires a heap allocation which I'm trying to avoid with the C++ iterator.

A new function that returns `std::vector<Point2>` would be easy to implement and would skip the need to make a class in the first place, but a correct iterator would be better for performance.

File: src/libtcod/bresenham.h [hunk 3]
```diff
@@ -95,38 +97,86 @@ namespace tcod {
...
     This function is provisional.
  */
-inline bool bresenham_line(int xFrom, int yFrom, int xTo, int yTo, const std::function<bool(int, int)>& callback) {
-  auto line = BresenhamLine(xFrom, yFrom, xTo, yTo);
-  do {
-    if (!callback(xFrom, yFrom)) {
+inline bool bresenham_line(
+    std::array<int, 2> from, std::array<int, 2> to, const std::function<bool(std::array<int, 2>)>& callback) {
+  for (auto&& [x, y] : BresenhamLine(from, to)) {
+    if (!callback({x, y})) {
```
Comment (HexDecimal) [inline src/libtcod/bresenham.h:168]: While I mean to target C++17 and support structured bindings the libtcod library itself targets C++11 and is tested with C++14.  This means you can't use structured bindings in libtcod, only in your own C++17 code which uses libtcod.
```suggestion
  for (auto&& xy : BresenhamLine(from, to)) {
    if (!callback(xy)) {
```

Comment (sensualcoder) [reply src/libtcod/bresenham.h:168]: I saw the warnings about C++17 too, my bad. Easy enough to fix up though.

**Comment by sensualcoder** [comment]
The vector heap allocation occurred to me when I was making that change, too. Hmm, ok, I'll have to go back to the drawing board on that implementation. Maybe I should break out the different pieces of `TCOD_bresenham_data_t and maybe move the step calculation into a separate function or lambda that can then calculate all the line points based on a step count or something like that.

**Comment by HexDecimal** [comment]
I'll probably continue to work on this class after you're done.  Until then step-count seems like a reliable way to track the range of the iterator without having to rewrite any of the Bresenham logic.

**Comment by sensualcoder** [comment]
I still need to figure out how to change it so it doesn't clip the endpoint.

**Comment by sensualcoder** [comment]
I just realized I have been trying to implement a RandomAccessIterator rather than a BidirectionalIterator, I'll rework that.

**Comment by sensualcoder** [comment]
Disregard my previous comment, I think the current implementation is set up correctly, except I think I need to invert the `step`, `delta`, and `e` values inside `TCOD_bresenham_data_t` when decrementing.

**Comment by HexDecimal** [comment]
Stepping backwards through a Bresenham line and reversing a Bresenham line are unfortunately not the same thing.  Bresenham lines are not symmetric on their own so I'd be worried that the reversed line will step through different positions than the forward line.  This would need testing to confirm it works correctly.

For now I'd be okay with a working forward-only iterator.

**Comment by sensualcoder** [comment]
I'm pretty sure I was overthinking that part actually, I'll try it a different way. The `step` method and forward iterator seem to work well currently.

**Comment by sensualcoder** [comment]
I think you're right about a reversed line creating a slightly different line, even if you were to init a new line with swapped `orig` and `dest` points. After researching and mulling if over I haven't been able to figure out a good solution for decrementing, but I may come back to it later.

**Review by HexDecimal — APPROVED** [review]
Looks good. With the forward iterator done the new `bresenham_line` function is no longer needed since any kind of callback-based or vector-based function can be implemented trivially from the iterator itself.

**Pull request merged**